### PR TITLE
Front-end tooling upgrade

### PR DESCRIPTION
**Pull request by thibaudcolas**
This PR does quite a bit more than I was intending to initially – the main reason for this is to simplify the upgrade of dependencies that are often interdependent. Headline changes are:

- Switching all Gulp usage to Webpack, in particular Sass -> CSS processing
- Updating to our latest Stylelint and ESlint configurations, which are compatible with Prettier. Including code cleanup.
- Switching Wagtail’s fonts to a system font stack
- Updating our version of Sass, updating deprecated patterns.

## Gulp to Webpack

The changes here consisted of:

- Moving the Sass files to be Webpack entries. 
- Moving all other asset files to be processed by either the copy plugin, or loaders

As part of this I chose to also remove our web font files and switch over to a system font stack (#7724). This is still WIP, but I believe should be merge-able as-is nonetheless, and we’ll be working on finishing that switch as soon as this PR is merged.

The main reason for doing this switch in this PR is that it felt silly to spend time reworking a font loading setup I knew would be going away in a few weeks. I’ve only done the cleanup of the existing fonts – we’ll still need to spend proper time on this according to #7724.

## Stylelint & ESLint

The main drive for this is to allow us to start using Prettier – we need compatible configurations. Our Stylelint configuration is pretty up-to-date so there wasn’t much to do. I did change the configuration file format to make it easier to copy-paste rules between our shared configuration and any overrides here.

For ESLint, Wagtail’s setup was lagging behind our shared configuration a lot. I chose to keep code changes to a minimum, and we’ll revisit which rules to enforce in a subsequent PR.

For our stylesheets – it’s worth noting there was a lot of duplicate code flagged. I’ve cleaned up only the instances I thought were 100% safe.

## Wagtail’s font stack

See the notes above in Gulp to Webpack.

## Sass version

The bulk of the changes here have been implemented with the automated [`sass-migrator`](https://sass-lang.com/documentation/cli/migrator), and will be outputting the exact same CSS.

There is a remaining deprecation note, coming from code we’re importing from Draftail. I’ll release an update to that package at some point, as part of which I’ll address this.

---

* [x] Do the tests still pass? (https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
* [x] Does the code comply with the style guide? (Run `make lint` from the Wagtail root)
* ~~For Python changes: Have you added tests to cover the new/fixed behaviour?~~
* [x] For front-end changes: Did you test on all of [Wagtail’s supported environments](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)?
    * **Please list the exact browser and operating system versions you tested**.
    * **Please list which assistive technologies you tested**.
* ~~For new features: Has the documentation been updated accordingly?~~

The main way I’ve tested this is with the visual regression testing suite of https://github.com/thibaudcolas/wagtail-tooling. Additionally I’ve done the following cross-browser testing: Safari 15.2 macOS, Chrome 97 macOS, Firefox 95 macOS, Edge 97 Windows 10, Firefox 91 ESR Windows 11, iPadOS Safari 13, iOS Safari 15, Chrome Android 10.

Additionally I tested the build tools work on Windows via CircleCI: https://app.circleci.com/pipelines/github/thibaudcolas/wagtail-fork/143/workflows/9d6052b3-fbea-4e47-9edb-02c5457c3e56/jobs/195. I’ll add this directly to Wagtail in a separate PR once #6790 is merged.

### CODE REVIEW

File: .eslintrc.js
Comment (thibaudcolas) [inline .eslintrc.js:2]: For everything in this list, we’ll need follow-up work to either disable the rule in our shared configuration, or refactor the code and enforce it here.

File: .eslintrc.js
Comment (thibaudcolas) [inline .eslintrc.js:20]: We weren’t using Webpack-style resolution anywhere. I think that’s a good thing.

File: client/.storybook/main.js
Comment (thibaudcolas) [inline client/.storybook/main.js:8]: I’ll be doing much more work with Storybook in a subsequent PR – just removing those for now as they’re largely unused currently.

File: client/src/components/StreamField/scss/_variables.scss
Comment (thibaudcolas) [inline client/src/components/StreamField/scss/_variables.scss:34]: This will be revisited in a subsequent PR – I just did the bare minimum so we could proceed with the removal of the existing font files.

File: docs/contributing/css_guidelines.rst
Comment (thibaudcolas) [inline docs/contributing/css_guidelines.rst:7]: I have removed this because those are essentially the same instructions as other front-end build tools now.

File: package.json
Comment (thibaudcolas) [inline package.json:98]: This is a dependency that was there already (used to format dates in comments only), it was just not declared at the top level. It’s completely inappropriate for Wagtail, as the file size is enormous for the value we’re getting from this. I would like us to remove this in a follow-up issue.

File: package.json
Comment (thibaudcolas) [inline package.json:55]: I chose not to upgrade to Storybook v6.4, as I’ll be spending much more time with this in a subsequent PR.

File: client/webpack.config.js
Comment (thibaudcolas) [inline client/webpack.config.js:13]: This is the same as before, just hopefully a bit easier to parse.

File: .eslintrc.js
Comment (lb-) [inline .eslintrc.js:55]: Should we put the typescript recommended into the ahared config? (Even as a seperate export maybe)? Not critical now though.

File: .eslintrc.js
Comment (lb-) [reply .eslintrc.js:20]: Nice.

File: .stylelintrc.js
Comment (lb-) [inline .stylelintrc.js:2]: I don't think this comment is necessary, most people reviewing this file would understand how the styelint extends system works.

File: .stylelintrc.js
Comment (lb-) [inline .stylelintrc.js:8]: Can this ignore rule be scoped go a subset of files only?
This way these issues won't creep into the code while we work on the font removal. 

https://stylelint.io/user-guide/configure/#overrides

File: .stylelintrc.js
Comment (lb-) [inline .stylelintrc.js:6]: Can this rule be scoped to anything outside of client only maybe? Or is there also 'legacy' code there?

https://stylelint.io/user-guide/configure/#overrides

File: client/scss/components/_indicator.scss
Comment (lb-) [inline client/scss/components/_indicator.scss:46]: Should this use math.div?

File: wagtail/documents/static_src/wagtaildocs/scss/add-multiple.scss
Comment (lb-) [inline wagtail/documents/static_src/wagtaildocs/scss/add-multiple.scss:2]: Is there any way to avoid these changes to relative imports? Just makes this code much harder to refactor later.

File: wagtail/admin/static_src/wagtailadmin/scss/userbar.scss
Comment (lb-) [inline wagtail/admin/static_src/wagtailadmin/scss/userbar.scss:309]: Also here - should math.div be used?

File: package.json
Comment (lb-) [inline package.json:125]: It would be nice if we didn't have to declare the ext here. Is that a performance decision? Is there any way to do this in the configuration? 
Do we currently lint JS in html files ( I don't think so but checking)

Comment (lb-) [reply package.json:125]: Do we need a script to clear any already built files? I know that sometimes this gets people stuck - me included.

File: package.json
Comment (lb-) [reply package.json:98]: Can an issue be raised from this?
Additional to that we may want to add a future ticket to limit the build filesize for things like this to get flagged early.

**Review by lb- — CHANGES_REQUESTED** [review]
Overall this looks good, I have a few additional questions and suggestions though. 

1. Can this PR be split into a. Eslint, b. Styelint, c. Fonts, d. Gulp removal - OR be rebased into four ish commits that are logically grouped. Just so that when this is in the main trunk it is much easier to follow. 
2. Has this been tested on windows? I can't see anything that should be a windows issue but that has caught us up in the past when changing build pipeline. 
3. What is the reasoning of removing the named import for helpers - I think that having to do deep parent nested import paths is a pain to do and easily breaks. 
```
@import 'wagtailadmin/scss/helpers';
// Replaced with -
@import '../../../../../../client/scss/settings';
@import '../../../../../../client/scss/tools';
```

This is amazing work Thibaud, thanks so much for your great skills in this area. I will take some time to run locally also and give any further feedback from there.

**Comment by lb-** [comment]
Not sure if I want to ask but should we update nvmrc to node 16? or leave that to a separate PR?

File: package.json
Comment (thibaudcolas) [reply package.json:98]: Done: #https://github.com/wagtail/wagtail/issues/7904. This will be enforced by [no-extraneous-dependencies](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-extraneous-dependencies.md) once we have our ESLint configuration upgrade merged :)

File: .eslintrc.js
Comment (thibaudcolas) [reply .eslintrc.js:55]: Yep, I like the idea. I’ve made a separate issue for it here: https://github.com/wagtail/eslint-config-wagtail/issues/17.

File: .stylelintrc.js
Comment (thibaudcolas) [reply .stylelintrc.js:8]: I’ll switch all legacy icon font usage to a Sass variable, so we can enforce this rule everywhere.

File: client/scss/components/_indicator.scss
Comment (thibaudcolas) [reply client/scss/components/_indicator.scss:46]: Indeed, I’ll refactor. The `* 0.5` comes from [sass-migrator](https://sass-lang.com/documentation/cli/migrator).

File: package.json
Comment (thibaudcolas) [reply package.json:125]: Re `--ext`, I believe this is [needed](https://eslint.org/docs/user-guide/command-line-interface#--ext) so `ESLint` will pick those files up when used with a folder as the target (`./`). See also https://typescript-eslint.io/docs/linting/#running-eslint.

I don’t think there is a way to do this via the configuration. `overrides` could help with this as mentioned in the docs, but I don’t think they’re intended to be used _just_ for the purpose of specifying which files to lint.

---

Re clearing built files, I’d be keen for us to try Webpack 5’s `output.clean: true` setting, and-or completely change how we output files so we no longer have this mess of 20 different output folders – but I’d rather do this in a follow-up PR.

**Comment by thibaudcolas** [comment]
@lb- good question – I’ll proceed with the changes you suggested, then take on the Node 16 + npm v8 upgrade, and see whether it makes sense in the same PR or separate.

I’m a bit hesitant because the npm v6 -> v8 upgrade is going to cause quite a bit of pain for us, but perhaps this is indeed worth doing at the same time.

**Comment by thibaudcolas** [comment]
I forgot two points:

## Windows support

I tested this with CircleCI’s Windows executor, in powershell: https://github.com/thibaudcolas/wagtail-fork/blob/d91344e88fca63681889ee6a358fe46d6feb653a/.circleci/config.yml#L63-L77, https://app.circleci.com/pipelines/github/thibaudcolas/wagtail-fork/143/workflows/9d6052b3-fbea-4e47-9edb-02c5457c3e56/jobs/195

I would like to add this to Wagtail’s CircleCI configuration, but only once #6790 is merged.

## wagtailadmin/scss/helpers removal

Yes, I’m not a fan of the change either. The previous style was only possible do to how `gulp-dart-sass` resolved imports, so I had to change this as part of the switch to Webpack. I’d like to revisit this – ideally I’d like us to have much fewer styles per-app like we do at the moment (I think about 90% of those styles are badly structured), so I’ve left the imports in this intermediary sad state for now.

**Comment by thibaudcolas** [comment]
@lb- updated with the requested changes, 6 commits including Node 16 + npm v8.

Also needed:

- https://github.com/wagtail/vagrant-wagtail-develop/pull/31
- https://github.com/wagtail/docker-wagtail-develop/pull/39

File: package.json
Comment (lb-) [reply package.json:119]: All makes sense, thanks for clarifying.

**Review by lb- — APPROVED** [review]
Looking great @thibaudcolas 

I just ran the fix command and it changed one file `wagtail/contrib/modeladmin/static_src/wagtailmodeladmin/js/prepopulate.js` (just a global comment). I will push that change to your branch if that is ok.

Beyond that, I cannot see anything else to fix, looks like we need to update our draftail package (especially with the future dart-scss changes coming) but I am sure that is on your roadmap elsewhere.

I have tested all the main package scripts and cannot see any further issues.

**Comment by thibaudcolas** [comment]
W00t!

**Pull request merged**
